### ISSUE TITLE
Duplicate Component Id resolution algorithm can cause upgrade issues

### ISSUE THREAD

**Issue by russcam** [oleg-shilo/wixsharp#203]
Component Guid determines the uniqueness of a component across product versions. 

WixSharp's default Guid generation uses a consistent hashing algorithm to generate consistent Guids for a given `<Component>` based on its `Id`, and for a `<Component>` that contains a `<File>`, the `Id` will be  the `Id` of the `<File>` prefixed with `Component.`. 

WixSharp avoids creating `<File>` elements with the same `Id` by appending an incrementing counter to the `Id` when it has seen that `Id` before e.g.

```xml
<Component 
    Id="Component.plugin_descriptor.properties" 
    Guid="daaa2cba-a1ed-4f29-afbf-547893a03ecb" Win64="yes">
    <File 
        Id="plugin_descriptor.properties" 
        Source="..\build\bar\plugin-descriptor.properties" />
</Component>
```
for the first `<File>` with file name `plugin_descriptor.properties` in one directory then

```xml
<Component 
    Id="Component.plugin_descriptor.properties.1" 
    Guid="daaa2cba-a1ed-4f29-afbf-5478fec21303" Win64="yes">
    <File 
        Id="plugin_descriptor.properties.1" 
        Source="..\build\foo\plugin-descriptor.properties" />
</Component>
```

for the second `<File>` with file name `plugin_descriptor.properties` in a different directory.

This approach is problematic however because a `<File>` Id in one MSI version e.g. `plugin_descriptor.properties`, may not be the same file path as the same `<File>` Id in another version. In the example above, version 2 of this MSI may not contain the file `"..\build\bar\plugin-descriptor.properties"`, so WixSharp will generate the Id `Component.plugin_descriptor.properties` for the file at `"..\build\foo\plugin-descriptor.properties"`. Since the MSI will use the `<Component>` Guid to determine if the `<Component>` is a shared component when performing an upgrade, it can lead to an upgrade removing files that it shouldn't whilst removing the old product version.

This can currently be worked around using `project.WixSourceGenerated += PatchWixSource;` to also include the directory as part of the `<File>` Id, then generate the wrapping `<Component>` Id and Guid from this new Id, but I think this ought to handled within WixSharp, probably as part of `WixEntity` Id getter (where the Id is derived from the Name): 

https://github.com/oleg-shilo/wixsharp/blob/9667cc9b3595f963048c25ca50055613a43f16b8/Source/src/WixSharp/WixEntity.cs#L288-L367

**Comment by oleg-shilo** [comment]
This issue may be related to #196

I think the ID based on the hash of the target (destination) path might be exactly what you need:
```C#
WixEntity.CustomIdAlgorithm =
    entity =>
    {
        if (entity is File file)
        {
            var target_path = project.GetTargetPathOf(file);
            var hash = target_path.GetHashCode32();

            // WiX does not allow '-' char in ID. So need to use `Math.Abs`
            return $"{target_path.PathGetFileName()}_{Math.Abs(hash)}";
        }
        return null;  // pass to default ID generator
    };
```
```xml
<Component Id="Component.Index.cshtml_1630848081" Guid="5495a725-b52b-4e09-96f6-14d3079908f6">
    <File Id="Index.cshtml_1630848081" Source="files\aaa\Index.cshtml" />
</Component>
```
The feature is so fresh that it is not described on Wiki yet.

**Comment by russcam** [comment]
G'Day back fron Sydney!

Yes it looks like the same issue. I think the custom ID algorithm generator is good, but I think this should be part of the default id generation implementation?

Another useful extension may also be to allow [setting Guid to `"*"` for Components that support it](http://wixtoolset.org/documentation/manual/v3/xsd/wix/component.html)?

**Comment by oleg-shilo** [comment]
> ... but I think this should be part of the default id generation implementation?

Absolutely. This is the intention. 

The feature is just 4 days old and I didn't want to rush it into being default until the impact is properly assessed. The idea is:
1. Implement custom id-generation algorithm. _Done_
2. Implement hashed-based id as an embedded feature.  _In progress_ 
3. Transition it into a default mode. _Planing_

Point 3 is the most difficult. The hash is calculated for the target path, which cannot be discovered by analyzing the `WixEntity` alone but only the `Project` object.  

However this breaks the current Id-generation paradigm based on the algorithm having an atomic nature and being completely decoupled from any deployment logic. I do prefer to keep this paradigm untouched that's why to move forward it will be required to injecting project context into the algorithm prior the build. It's all doable but requires careful crafting as side effect can easily be triggered (e.g. when `await project.BuildAsync();`)   

As for `"*"` for components, I am thrilled to see this feature. I wasn't aware of it. Finally, WiX has accepted that insisting on verbosity for the sake of ultimate explicitly truly damages the user experience. 

Effectively WixSharp does `"*"` for everything, not only for the components. I'll see if there are the benefits of integrating component's "*".

**Comment by oleg-shilo** [comment]
Please have a look at the `Release v1.4.11.0 (RC)` release. 

This is how you can enable the hash algorithm:

```C#
var project = new Project("MyProduct",
                  new Dir(@"%ProgramFiles%\My Company\My Product",
                      new File(@"Files\Bin\MyApp.exe"),
                      new Dir(@"Docs\Manual",
                          new File(@"Files\Docs\Manual.txt"))));

Compiler.AutoGeneration.LegacyDefaultIdAlgorithm = false;
```

Please not that it is a pre-release and `LegacyDefaultIdAlgorithm` is set to `true` by default. It will be changed in the future releases to `false` but only after the WixSharp users get used to the new approach. BTW some users store the produced WXS under source control thus the switch to the new algorithm will immediately trigger the changes in the codebase.

**Comment by russcam** [comment]
Thanks for moving on this so quick @oleg-shilo! Agree that suddenly changing the behaviour could be very disruptive to existing users; would this warrant a major version bump when the release is GA?

**Comment by oleg-shilo** [comment]
> ...would this warrant a major version bump...

No, my current plan is to let it stay as pre-release (v1.4.11) for a month or so and then default it to the new algorithm and release with a _minor version_ increase ( v1.5.0). After all, strictly speaking the new approach is still backwards compatible. 

Though, I am trying to proceed cautiously. Thus in order to advertise the new feature and also see if the change of the default algorithm would be welcome I have logged a new issue "Community feedback needed... #204". You can also vote there.

**Comment by oleg-shilo** [comment]
The very latest release still has the new algorithm disabled by default. 

Though, it has been two month since the introduction of this new feature and there were no objection form the users (primary objective of voting).

Thus the feature is already enabled by default in the codebase (0fc5887f9b491d6ebd646942211a672081bea75d) and the very next release will deliver it to the public.
